client/src/main/java/io/pravega/client/stream/impl/PositionImpl.java

```diff
@@ -196,4 +230,20 @@ public static Position fromBytes(ByteBuffer buff) {
         return SERIALIZER.deserialize(new ByteArraySegment(buff));
     }
 
+    private synchronized void applySegmentOffsetUpdatesIfNeeded() {
+        // No updates, so nothing to do.
+        if (this.updatesToSegmentOffsets == null) {
+            return;
+        }
+        // We are going to modify ownedSegments to match the segment offsets at the time this event was read, so copying
+        // the map of ownedSegments to do not impact to the map referenced by other PositionImpl objects.
+        this.ownedSegments.set(new HashMap<>(ownedSegments.get()));
+        // Apply all the Segment offset updates up to the point in which this event was read.
+        for (int i = 0; i < this.version; i++) {
+            this.ownedSegments.get().put(this.updatesToSegmentOffsets.get(i).getKey(), this.updatesToSegmentOffsets.get(i).getValue());
```

Comment:
Thanks for the comment @shrids. I assume that your concern is about the possibility of getting a `ConcurrentModificationException` while looping over a list that could be subject to concurrent appends (totally valid concern). To test whether this is thread safe or not, I have tested the following:
1) Thread is continuously adding elements on a list.
2) Another thread is continuously iterating the same list.

I have tried 2 variants for thread 2:
- Looping over the entire list:
```
for (Integer element: elements) { //do X on element }
```
- Emulating what we do in this PR (take a point in the list before the end and iterate up to that point):
```
for (int j = 0; j < elements.size() / 2; j++) {//do X on element}
```
For case 1, I consistently get `ConcurrentModificationException` on every execution. However, after many executions of this experiments, I have not seen any occurrence of this problem for case 2. I accept that this may not be a usual pattern, but is the most efficient approach I could come up so far and it seems to be thread safe.

Reply:
With the current implementation of ArrayList i do not see the possibility of an errror, given that we guarantee only one thread is performing the add. The `resize` impl of the ArrayList would not cause an error here too. It is just that the pattern is unconventional and I have always avoided it.